### ISSUE TITLE
unmask CPSR A bit during fast SMC

### ISSUE THREAD

**Issue by milunt** [OP-TEE/optee_os#933]
```diff
sm_set_sec_smc_entry (){
if(FASTSMC) {
      sec_ctx->mon_lr = (uint32_t)&sm_entry_vector->fast_smc_entry
 +  sec_ctx->mon_spsr =   I |  F | cpsr_mode_svc  
}
```

while exiting to sec `rfefd`  will run in `sec_svc`  mode with `I` and `F` bits set but `A` unmasked.
Do you think this is the ideal condition to run fast_smcs.

Or what would be a reason to keep abort exceptions masked during fast smc handling like it is now ?

Thanks
Sandeep

**Comment by jenswi-linaro** [comment]
What is it that you'd like to do with the asynchronous abort?

**Comment by milunt** [comment]
Just to allow taking async aborts all the time.  why not. 

okay, how I got into this is:
 run hello_world
 then do any fast smc
 I get  ERR TEE-CORE:_assert_log:38: Assertion 'thread_get_exceptions() == THREAD_EXCP_ALL' failed at core/arch/arm/kernel/thread.c:537

So 
  1. first of all I think we are restoring to a saved mon_spsr to run fast smc handler.
      which does not seem right since it is always a fresh run we should be just deciding the CPSR state.
      so with that patch I am ok.  
  2. Then why to expect async abort to be masked in secure SVC mode where we are running fast smc.     

  3.  also how about masking async abort on exit of stdsmc  as follows.

diff --git a/core/arch/arm/kernel/thread_a32.S b/core/arch/arm/kernel/thread_a32.S
index 0f32fb9..09ce322 100644
--- a/core/arch/arm/kernel/thread_a32.S
+++ b/core/arch/arm/kernel/thread_a32.S
@@ -308,7 +308,7 @@ UNWIND(     .cantunwind)
        add     sp, #(4 * 4)

        /* Disable interrupts before switching to temporary stack */
-       cpsid   if
+       cpsid   aif
        bl      thread_get_tmp_sp
        mov     sp, r0

**Comment by jforissier** [comment]
Please use a code block when you copy/pase some code or diff patch. Otherwise formatting is broken and it's hard to follow. To do this, either indent your code with 4 spaces or a tab, or use three backticks like so:

    ```
    Some code
    ```
(if you're pasting stuff it's easier to use the backtick method).

**Comment by jenswi-linaro** [comment]
If you're not doing anything useful when receiving it, why bother. I think it's more useful for normal world to log or whatever when it happens as it's likely that normal world triggered it from the beginning.

The problem with unmasking async abort that while we're on the temporary stack is that it may be used by some exception handlers or that stacks for exception handlers isn't properly set up yet.

**Comment by milunt** [comment]
thx Jforissier ,let me try this time ;) 

Agree that there are windows where you can not reliably take an exception. but abort is something you can not avoid just because we are in secure world. So we are discussing two topics here. One 
should we take async abort while handling in fast_smc or not. 

The first topic is "should I override a sane state value to sec_ctx->mon_spsr while entering fast_smc."
Otherwise you know the issue as described earlier.   run any app.  mon_spsr  becomes  0x600000d3.
then  fast_smc  will abort as A is not masked.
I purpose to do the following. 

` void sm_set_sec_smc_entry(const struct sm_reg_r0_to_r3 *regs)
 {
        struct sm_sec_ctx *sec_ctx = sm_get_sec_ctx();

-       if (OPTEE_SMC_IS_FAST_CALL(regs->r0))
+       if (OPTEE_SMC_IS_FAST_CALL(regs->r0)) {
                sec_ctx->mon_lr = (uint32_t)&sm_entry_vector->fast_smc_entry;
+               sec_ctx->mon_spsr |= (CPSR_MODE_SVC | ARM32_CPSR_FIA);
+       }
        else
                sec_ctx->mon_lr = (uint32_t)&sm_entry_vector->std_smc_entry;
+
 }
`

**Comment by jforissier** [comment]
@milunt almost there ;) you need *3* backticks, not just one.

**Comment by jenswi-linaro** [comment]
Sorry, it's very hard to parse the code.

**Comment by milunt** [comment]
Ahh I missed there is a preview option .  so here is take 2..
```
 void sm_set_sec_smc_entry(const struct sm_reg_r0_to_r3 *regs)
 {
        struct sm_sec_ctx *sec_ctx = sm_get_sec_ctx();

-       if (OPTEE_SMC_IS_FAST_CALL(regs->r0))
+       if (OPTEE_SMC_IS_FAST_CALL(regs->r0)) {
                sec_ctx->mon_lr = (uint32_t)&sm_entry_vector->fast_smc_entry;
+               sec_ctx->mon_spsr |= (CPSR_MODE_SVC | ARM32_CPSR_FIA);
+       }
        else
                sec_ctx->mon_lr = (uint32_t)&sm_entry_vector->std_smc_entry;
+
 }
```

**Comment by milunt** [comment]
```
 void sm_set_sec_smc_entry(const struct sm_reg_r0_to_r3 *regs)
 {
        struct sm_sec_ctx *sec_ctx = sm_get_sec_ctx();
+       uint32_t fast_smc_cpsr;

-       if (OPTEE_SMC_IS_FAST_CALL(regs->r0))
+       if (OPTEE_SMC_IS_FAST_CALL(regs->r0)) {
                sec_ctx->mon_lr = (uint32_t)&sm_entry_vector->fast_smc_entry;
+               fast_smc_cpsr = sec_ctx->mon_spsr;
+               fast_smc_cpsr &= CPSR_MODE_MASK;
+               fast_smc_cpsr |= (CPSR_MODE_SVC | ARM32_CPSR_FIA);
+               if (sec_ctx->mon_lr & 1)
+                       fast_smc_cpsr |= CPSR_T;
+               sec_ctx->mon_spsr = fast_smc_cpsr;
+       }
        else
                sec_ctx->mon_lr = (uint32_t)&sm_entry_vector->std_smc_entry;
+
 }

```

**Comment by jenswi-linaro** [comment]
Aborts in a fast call are fatal so it doesn't matter much if we destroy the stack while taking the exception.
When OP-TEE does the SMC to switch into the monitor it's supposed to leave CPSR in the state it need when returning later so there shouldn't be a need to do anything but save and restore SPSR in the secure monitor.

Before you dig any deeper in this, what is it with the asynchronous aborts that you need to handle during a fast call?

**Comment by milunt** [comment]
I am now keeping the need for abort in fast_call aside. That might at best give some debug clue in case of abort during a fast_call handling code. nothing much.

I see in some cases the CPSR.A  remains unmasked while returning  thread_std_smc_entry .
not sure why. I will check that too.
ex:  hello_world
So as expected in fast call entry it get CPSR.A bit unmasked . which is unintended.

Could you please check my second post where I have purposed another patch to mask 
'aif' after returning from  __thread_std_smc_entry.

```diff
diff --git a/core/arch/arm/kernel/thread_a32.S b/core/arch/arm/kernel/thread_a32.S
index 0f32fb9..09ce322 100644
--- a/core/arch/arm/kernel/thread_a32.S
+++ b/core/arch/arm/kernel/thread_a32.S
@@ -308,7 +308,7 @@ UNWIND(     .cantunwind)
        add     sp, #(4 * 4)

        /* Disable interrupts before switching to temporary stack */
-       cpsid   if
+       cpsid   aif
        bl      thread_get_tmp_sp
        mov     sp, r0
```

**Comment by jenswi-linaro** [comment]
That patch is already applied on master, but thanks any way. :-)

**Comment by milunt** [comment]
I see.  Thanks a lot.
